### PR TITLE
Controlling select_speaker using Graph Modelling Language

### PR DESCRIPTION
**Pull request by joshkyh**
I suggest @afourney and @thinkall review this work because of their recent contributions to incorporate modes of select_speaker in #688.

## Why are these changes needed?

This notebook shows how we can use an expressive graph to control select_speaker. Instead of hard-coding complex logic to override `select_speaker`, this example shows how we can input a directed graph that dictates the transition paths between agents. If desired, we can also extend GroupChat `select_speaker` mode to "graph" in the near future (extending the work in #688) in another PR. I believe this is a necessary step towards having more agents working together, since the number of transition pairs increases rapidly (N Choose 2).

image

## Related issue number

#485 
#438 in terms of dividing the problem.

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by afourney** [comment]
Thanks. This looks really compelling. I need to dig into it more.

One thing I noticed is that it selects the next speaker randomly from among the candidate outgoing edges (or did I misread that?). I would love to see this instead use some variation of the current GroupChat prompt (or similar) to select the next speaker based on the conversations.

**Comment by joshkyh** [comment]
> Thanks. This looks really compelling. I need to dig into it more.
> 
> One thing I noticed is that it selects the next speaker randomly from among the candidate outgoing edges (or did I misread that?). I would love to see this instead use some variation of the current GroupChat prompt (or similar) to select the next speaker based on the conversations.

Hi @afourney,

Glad you find it useful. A partial yes, if `suggested_next` by the previous speaker is a legal transition, then that transition would happen. `suggested_next` is not random, it is suggested by the previous speaker. However, if it is not a legal transition, we would pick a random agent from the legal transitions `eligible_speakers` as a catch-all. Yes, we can potentially insert the use of an LLM to pick an agent from the `eligible_speakers` instead, like the existing GroupChat.

Please feel free to review them when you get some time. I'll act on them tomorrow. :)

image

**Comment by BeibinLi** [comment]
Fantastic idea! I have thought about something similar earlier, but your implementation is much cleaner. I will take a deeper look later.

**Review by thinkall — COMMENTED** [review]
Thanks @joshkyh , I love it. Allowing selecting speaker based on the complex network would push the group chat to a higher level, enabling a more meaningful large group chat.

Do you only want to show it as a notebook example? I believe it worth to add into the code.

**Comment by joshkyh** [comment]
Hi @thinkall and @BeibinLi, 

Thank you for the kind comments. I'm glad you found it useful enough to be added to the codebase. I can do that but on a part-time basis. Can we first merge this notebook into the repo so that people can refer to it? I can then work on the main codebase, and refactor some of the `GroupChat` code, perhaps adding a new class `GraphGroupChat` to include a `graph` mode.

I am OK with the Team wanting to take this internally and develop it too. Just let me know your preferences. :)

@afourney, I have extended the notebook with a snippet from `GroupChat` that uses the LLM to pick from eligible speakers when the transition is not legal, and before the random catch-all assignment.

**Review by afourney — APPROVED** [review]

**Comment by afourney** [comment]
I approved the notebook. I'd like to get that class into some Testbed scenarios when I get a chance, to see what it can do.

File: notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture --no-stderr\n",
+    "# %pip install pyautogen~=0.2.0b5\n",
+    "# pip install networkX~=3.2.1\n",
+    "# pip install matplotlib~=3.8.1"
```
Comment (sonichi) [inline notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb:43]:
```suggestion
    "# %pip install pyautogen~=0.2.0b6\n",
    "%pip install networkX~=3.2.1\n",
    "%pip install matplotlib~=3.8.1"
```
uncomment the pip install for enabling the test

File: notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "0.2.0b5\n"
```
Comment (sonichi) [inline notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb:55]:
```suggestion
      "0.2.0b6\n"
```

File: notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb [hunk 3]
```diff
@@ -0,0 +1,837 @@
...
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Motivation\n",
+    "\n",
+    "\n",
+    "The current GroupChat class allows transition to any agent (without or without the decision of LLM), some use case might demand for more control over transition. A graph is a possible way to control the transition paths, where each node represent an agent and each directed edge represent possible transition path. Let's ilustrate the current transition paths for a GroupChat with five agents."
```
Comment (sonichi) [inline notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb:80]:
```suggestion
    "The current GroupChat class allows transition to any agent (without or without the decision of LLM), some use case might demand for more control over transition. A graph is a possible way to control the transition paths, where each node represents an agent and each directed edge represent possible transition path. Let's illustrate the current transition paths for a GroupChat with five agents."
```

File: notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb [hunk 4]
```diff
@@ -0,0 +1,837 @@
...
+    "        'base_url': '<your Azure OpenAI API base here>',\n",
+    "        'api_type': 'azure',\n",
+    "        'api_version': '2023-06-01-preview',\n",
+    "    },\n",
+    "]\n",
+    "```\n",
+    "\n",
+    "If you open this notebook in colab, you can upload your files by clicking the file icon on the left panel and then choosing \"upload file\" icon.\n",
+    "\n",
+    "You can set the value of config_list in other ways you prefer, e.g., loading from a YAML file."
```
Comment (sonichi) [inline notebook/agentchat_graph_modelling_language_using_select_speaker.ipynb:309]: Could you update this section following most of the other notebooks?

**Review by sonichi — COMMENTED** [review]
Would you like to add a link to this notebook in the website? There are two pages to update.

**Comment by joshkyh** [comment]
> Would you like to add a link to this notebook in the website? There are two pages to update.

Thanks for the review, @sonichi. I have acted on the four feedback points and added the link to this notebook in two pages.

**Comment by Friedensfreak** [comment]
Hi @joshkyh what is `secret` for each agent in the graph modelling. And how is this useful?

**Comment by Morriz** [comment]
will this be codified? Sharing code override examples is not the way forward imo.

**Comment by sonichi** [comment]
See #743 and related PRs.

**Pull request merged**